### DISCUSSION TITLE
Re-write streamdeck-linux-gui

### DISCUSSION THREAD

**Discussion by coolapso** [streamdeck-linux-gui/streamdeck-linux-gui#179] [Polls]
Python is not the primary language for most of our core developers and maintainers, myself included. This poses several challenges and diminishes the satisfaction we derive from maintaining an application that aims to be helpful to as many people as possible.

Currently, we face several significant issues:

    None of us find joy in writing and maintaining such a complex codebase in Python.
    Implementing new features is challenging.
    Developing a plugin system is difficult.
    Reviewing pull requests is cumbersome.
    Maintaining dependencies, especially when interfacing with end-users, is painfully difficult.

In light of these challenges, we are considering the option of rewriting the entire application, including any necessary libraries.

Our goals for this rewrite are:

    Establishing an easier-to-maintain codebase by redesigning the application from the ground up before commencing coding.
    Simplifying dependency management.
    Striving to adhere to modern software development best practices, including improving decoupling between the core application APIs and the UI.
    Streamlining the process of building and shipping the application to end-users.

We acknowledge that rewriting an application is rarely advisable and will require significant time and effort. However, we feel stuck between a rock and a hard place, lacking enthusiasm to maintain something unfamiliar to us.

The floor is open for discussion and we value your input on this matter and welcome your support in this major endeavor! Please participate in the poll, with the languages the core maintainers are more comfortable with and sharing your thoughts on this proposal. 

Your feedback is crucial as we navigate this decision-making process. Thank you for your collaboration and understanding.

obs: Yes .. ChatGPT helped writing this, I am not a very good writer myself :D 
Obs2: This is not a decision making thread, the idea is to scout and gather the wider community feedback and thoughts around it!

**Comment by oparada1988** [comment]
I say yes, but I don't know which of the yes options to select because I don't know which language is better than the other.

But my answer is "yes"

**Comment by Quintar** [comment]
I'm not part of the core team or anything but I can do C++ and I'd say it's sufficently high level to be useable. That said, considering that we used Python we can almost use Bash or Power Script but that's not very portable, which I think should be a main concern.

**Comment by NicolasGoeddel** [comment]
I am fluent in Python and can not see the issues you are facing. Python has a very big community and you will find solutions to nearly every problem easily. A plugin system would also be easy imho because you can import external python code without compiling anything. You just need the right architecture to begin with.

So I would like to stay with Python.

But if you really want to switch, please don't use NodeJS or C. Use modern compilable languages like Rust, Go or the latest C++ version.

Btw. I did not know about the new releases it seems. I still have running version 2.0.15. It looks like I have to try version 4.1.0 now. :open_mouth:

**Reply by coolapso** [reply]
We are open to keep python as long as there's knowledge around and someone that wishes to stick around and help with it. i'd say the biggest hurdles are: 
* Application architecture
* Distributing the application to the end-user is terribly painful. most of the issues we stumble upon its because ppl don't know how to install, or have the wrong version of the dependencies, or do who knows what with their systems.  of course this issue will always exist but we can minimize it as much as possible.

**Reply by coolapso** [reply]
Also ... looking at the current code base ... 
* What would you do? 
* Are you willing to take a stab at it and help maintaining it?

**Reply by NicolasGoeddel** [reply]
With virtual environments you should not have any problems with dependencies, except there are on OS level. :thinking: 

I did not look deeply into the code yet. I can give it a try surely. But it will take a few days until I got time.

**Reply by ajbucci** [reply]
Our main source of pain has been Fedora which uses bleeding edge versions of Python, and I'm not sure about how much complexity would be added trying to manage python installs in the user's environment. I don't think any of us are particularly experienced with shipping Python apps that work across all distros. Certainly none of us have extensive experience with snap, flatpak, appimage etc which could all be good options. Perhaps there is another potential solution that we aren't aware of.

**Reply by NicolasGoeddel** [reply]
I see. Unfortunately I also don't have much experience with creating packages, except RPMs because I do that at work. I never created something like snap, flatpak or appimage. I have only experience in creating container images for docker or kubernetes and such things. :thinking:

**Reply by Quintar** [reply]
Well that are all things one can learn I'd say. I'm a fan of AppImages too.

**Reply by coolapso** [reply]
>I see. Unfortunately I also don't have much experience with creating packages, except RPMs because I do that at work. I never created something like snap, flatpak or appimage. I have only experience in creating container images for docker or kubernetes and such things. 🤔

Which is the perfect scenario for any python applications where you have full control over the environment they run at which is not the case for  the linux-streamdeck-gui

**Reply by ajbucci** [reply]
> Well that are all things one can learn I'd say. I'm a fan of AppImages too.

Sure, I've personally sunk tens of hours of my time looking into the 3 options, and have a minimally working flatpak build (though it appears another contributor has one too that may be in a better state than mine). However, the documentation for all 3 is lacking in my opinion. Additionally, at least for flatpak, changes are required to the app code itself to get everything working in the sandbox. The app itself only really works well with full system access so we're adding complexity to our build without making use of some of the main selling points for using those delivery methods.

**Reply by coolapso** [reply]
I've sunk my fair share of time too trying to dig into both snap and flatpack was the first time ever I've given up on trying to get something to work :sweat_smile:

**Reply by NicolasGoeddel** [reply]
For AppImages there seems to be a simple solution: https://github.com/niess/python-appimage
But since we need to create udev rules and direct access to devices I am not sure if it stays that easy.

**Reply by theorowlett** [reply]
> Our main source of pain has been Fedora which uses bleeding edge versions of Python, and I'm not sure about how much complexity would be added trying to manage python installs in the user's environment. I don't think any of us are particularly experienced with shipping Python apps that work across all distros. Certainly none of us have extensive experience with snap, flatpak, appimage etc which could all be good options. Perhaps there is another potential solution that we aren't aware of.

I think pipenv is the way forward to handle dependencies and python versioning. I just now stumbled across this repo (I ordered a streamdeck plus yesterday) and I'll look into contributing where I can.

**Comment by v4u6h4n** [comment]
My coding knowledge maxes out with bash scripting, so I don't have much to add besides feeling passionate about a project is essential, and that once I'm making a little more money streaming I'd love to financially support this project, it's a game changer for me :-)

**Comment by Danie10** [comment]
It is really up to the devs as to what they feel most comfortable with, and can do their best work. If we had a few active Python devs then it is fine to stay with Python, but otherwise a change is probably a good thing. As long as it can still support current functionality and platforms as the users are using it.

Just hope it does not mean registering yet a new Github project as we already rebuilt all the issues a d discussions here.

**Reply by coolapso** [reply]
will never involve a new github project :)

**Comment by coolapso** [comment]
For those wondering and around this discussion, the solution presented itself: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/203
